src/storage/src/hummock/iterator/skip_watermark.rs

```diff
@@ -0,0 +1,382 @@
+                    Ordering::Equal => {
+                        match direction {
+                            WatermarkDirection::Ascending => {
+                                match inner_key.cmp(watermark.as_ref()) {
+                                    Ordering::Less => {
+                                        // The current key will be filtered by the watermark.
+                                        // Return true to further advance the key.
+                                        return true;
+                                    }
+                                    Ordering::Equal | Ordering::Greater => {
+                                        // The current key has passed the watermark.
+                                        // Advance the next watermark.
+                                        self.remain_watermarks.pop_front();
+                                        continue;
+                                    }
+                                }
+                            }
+                            WatermarkDirection::Descending => {
+                                return match inner_key.cmp(watermark.as_ref()) {
+                                    // Current key as not reached the watermark. Just return.
+                                    Ordering::Less | Ordering::Equal => false,
+                                    // Current key will be filtered by the watermark.
+                                    // Return true to further advance the key.
+                                    Ordering::Greater => true,
+                                };
+                            }
+                        }
+                    }
+                    Ordering::Greater => {
+                        return false;
+                    }
+                }
+            }
+        }
+        false
+    }
+
+    /// Advance the key until iterator invalid or the current key will not be filtered by the latest watermark.
+    /// Calling this method should ensure that the first remaining watermark has been advanced to the current key.
+    ///
+    /// Return a flag indicating whether should later advance the watermark.
+    async fn advance_key(&mut self) -> HummockResult<bool> {
+        if let Some((table_id, vnode, direction, watermark)) = self.remain_watermarks.front() {
+            while self.inner.is_valid() {
+                {
+                    let key = self.inner.key();
+                    let key_table_id = key.user_key.table_id;
+                    let (key_vnode, inner_key) = key.user_key.table_key.split_vnode();
+                    match (&key_table_id, &key_vnode).cmp(&(table_id, vnode)) {
+                        Ordering::Less => {
+                            return Ok(false);
+                        }
+                        Ordering::Equal => {
+                            if direction.filter_by_watermark(inner_key, watermark) {
+                                self.inner.next().await?;
```

Comment:
I am concern that we will loop over a large number of keys before exiting the loop, just like the previous issue with range tombstone. Should we call `seek` instead of `next`?
- WatermarkDirection::Ascending: seek to watermark
- WatermarkDirection::Descending: seek to next vnode